src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -203,66 +199,37 @@ public unsafe MsQuicConnection(QuicClientConnectionOptions options)
-                {
-                    if (listenerState.PendingConnections.TryRemove(state.Handle.DangerousGetHandle(), out MsQuicConnection? connection))
-                    {
-                        // Move connection from pending to Accept queue and hand it out.
-                        if (listenerState.AcceptConnectionQueue.Writer.TryWrite(connection))
-                        {
-                            return QUIC_STATUS_SUCCESS;
-                        }
-                        // Listener is closed
-                        connection.Dispose();
-                    }
-                }
-
-                return QUIC_STATUS_USER_CANCELED;
-            }
-            else
-            {
-                // Connected will already be true for connections accepted from a listener.
-                Debug.Assert(!Monitor.IsEntered(state));
-
-
-                Debug.Assert(state.Connection != null);
-                state.Connection._localEndPoint = MsQuicParameterHelpers.GetIPEndPointParam(MsQuicApi.Api, state.Handle, QUIC_PARAM_CONN_LOCAL_ADDRESS);
-                state.Connection.SetNegotiatedAlpn((IntPtr)connectionEvent.CONNECTED.NegotiatedAlpn, connectionEvent.CONNECTED.NegotiatedAlpnLength);
-                state.Connection = null;
+            Debug.Assert(state.Connection != null);
+            //state.Connection._remoteEndPoint = MsQuicParameterHelpers.GetIPEndPointParam(MsQuicApi.Api, state.Handle, QUIC_PARAM_CONN_REMOTE_ADDRESS);
+            state.Connection._localEndPoint = MsQuicParameterHelpers.GetIPEndPointParam(MsQuicApi.Api, state.Handle, QUIC_PARAM_CONN_LOCAL_ADDRESS);
+            state.Connection._negotiatedAlpnProtocol = new SslApplicationProtocol(new Span<byte>(connectionEvent.CONNECTED.NegotiatedAlpn, connectionEvent.CONNECTED.NegotiatedAlpnLength).ToArray());
+            state.Connection = null;
 
-                state.Connected = true;
-                state.ConnectTcs!.SetResult(QUIC_STATUS_SUCCESS);
-                state.ConnectTcs = null;
-            }
+            state.ConnectTcs.TrySetResult();
 
             return QUIC_STATUS_SUCCESS;
         }
 
         private static int HandleEventShutdownInitiatedByTransport(State state, ref QUIC_CONNECTION_EVENT connectionEvent)
         {
-            if (!state.Connected && state.ConnectTcs != null)
+            if (!state.ConnectTcs.IsCompleted)
             {
                 Debug.Assert(state.Connection != null);
                 state.Connection = null;
 
-                Exception ex = new MsQuicException(connectionEvent.SHUTDOWN_INITIATED_BY_TRANSPORT.Status, "Connection has been shutdown by transport");
-                state.ConnectTcs!.SetException(ExceptionDispatchInfo.SetCurrentStackTrace(ex));
-                state.ConnectTcs = null;
+                state.ConnectTcs.TrySetException(new MsQuicException(connectionEvent.SHUTDOWN_INITIATED_BY_TRANSPORT.Status, "Connection has been shutdown by transport"));
```

Comment:
Here and elsewhere, .resx.

Reply:
We should address this in #71432 which touches all the exceptions. I'm merely shuffling code here.